File: personnr.js
const personNr = (function () {
    let isValidBirthday = (day, month, year) => {

        const isLeapYear = (year % 4 === 0 && year % 100 !== 0) || (year % 400 === 0);

        const days = {
            1: 31,
            2: isLeapYear ? 29 : 28,
            3: 31,
            4: 30,
            5: 31,
            6: 30,
            7: 31,
            8: 31,
            9: 30,
            10: 31,
            11: 30,
            12: 31
        };

        if (year < 1900 || year > 2039) return false;
        if (month < 1 || month > 12) return false;

        if (day < 1 ||  day > days[month]) return false;

        return true;
    }

    let getRange = (year) => {
        let isBetween = (min, max, x) => x >= min && x < max;

        return isBetween(1900, 1999, year) ? [0, 499] :
            isBetween(2000, 2039, year) ? [500, 999] : [null, null];
    }

    let generatefrom = (dd, mm, yyyy) => {
        if (!dd ||  !mm ||  !yyyy) throw new Error('Expected a day, month and year with the following format: dd, mm, yyyy');
        if (!isValidBirthday(dd, mm, yyyy)) throw new Error('Expected format: dd, mm, yyyy');

        const [min, max] = getRange(yyyy);
        const [d, m, y] = [dd, mm, yyyy].map(n => n.toString());
        const result = [];
        
        [...Array(max - min + 1)]
        .map((u, i) => (min + i) < 10 ? '00' + (min + i) : (min + i) < 100 ? '0' + (min + i) : '' + (min + i))
            .forEach(n => {
                const k1 = 11 - ((3 * d[0] + 7 * d[1] + 6 * m[0] + 1 * m[1] + 8 * y[2] + 9 * y[3] + 4 * n[0] + 5 * n[1] + 2 * n[2]) % 11);
                const k2 = 11 - ((5 * d[0] + 4 * d[1] + 3 * m[0] + 2 * m[1] + 7 * y[2] + 6 * y[3] + 5 * n[0] + 4 * n[1] + 3 * n[2] + 2 * k1) % 11);
                if (k1 < 10 && k2 < 10) result.push(d + m + y[2] + y[3] + n + k1 + k2);
            });

        return result;
    }


    return {
        generatefrom
    };
})();



console.log(personNr.generatefrom(16, 10, 1990));

File: personnr.py
import re
from calendar import monthrange

def is_valid_date(d):
	if not re.match('^\d+$', d):
		raise ValueError('Not a valid date')
		
	dd, mm, yyyy = [int(i) for i in [d[:2], d[2:4], d[4:]]]
	
	if yyyy < 1855 or yyyy > 2039: 
		return False
	if mm < 1 or mm > 12: 
		return False
	if dd < 1 or dd > monthrange(yyyy, mm)[-1]: 
		return False
	
	return True
	
def get_range(year):
	if year >= 1855 and year <= 1899:
		return range(500, 750)
	if year >= 1900 and year <= 1999: 
		return range(500)
	if year >= 2000 and year <= 2039: 
		return range(500,1000)
		
	return range(0)	

def generate_national_id_numbers(d):
	if type(d) is not str or not is_valid_date(d):
		raise ValueError('Expected a valid date-string with the format: DDMMYYYY, but got: {}'.format(d))

	year, dt, result = int(d[4:]), [int(i) for i in d], []
	
	for i in get_range(year):
		n = list(map(int, list(str(i).zfill(3))))
		k1 = 11 - ((3*dt[0] + 7*dt[1] + 6*dt[2] + 1*dt[3] + 8*dt[-2] + 9*dt[-1] + 4*n[0] + 5*n[1] + 2*n[2]) % 11)
		k2 = 11 - ((5*dt[0] + 4*dt[1] + 3*dt[2] + 2*dt[3] + 7*dt[-2] + 6*dt[-1] + 5*n[0] + 4*n[1] + 3*n[2] + 2*k1) % 11)
		
		if k1 < 10 and k2 < 10: 
			result.append(''.join(list(map(str, dt[:4] + dt[6:] + n + [k1,k2]))))
			
	return result
	


print(generate_national_id_numbers('16101990'))